### ISSUE TITLE
icdrama website changes breaks parsing

### ISSUE THREAD

**Issue by jcheong-1** [aznhusband/aznhusband.github.io#43]
Hi Aznhusband,

Looks like changes to the icdrama site is breaking the functionality of this again. Would you kindly have a look whenever you have some time?

Thanks!

**Comment by tofufi** [comment]
same problem here

**Comment by 9kitsune** [comment]
Same here

**Comment by mugol** [comment]
It looks like something broke at mirror scrapping step. I think I found another way to grab the mirror links. The site is pretty wack with obscuring the video player itself though. I can look into this more over the weekend.

**Comment by aznhusband** [comment]
Yeah, Reddit's blowing up with this.  I'll start poking around with it, mugol, if you come up with anything, post in this bug, we'll use this bug to track progress.

**Comment by dfungfung** [comment]
So glad that I found this thread. Created an account to say that yes I have the same problem. Wish this would be fixed soon so I can keep my wife happy ;)

**Comment by aznhusband** [comment]
Wishing won't get it fixed, dfungfung.  We'll have to find the time to dig into the code, look at the updates to the website,and adjust all the scrapers.

**Comment by GhostlVlan** [comment]
Just wanted to say Thank You aznhusband and all the other developers working on maintaining this!  Super appreciative!

**Comment by dfungfung** [comment]
> Wishing won't get it fixed, dfungfung. We'll have to find the time to dig into the code, look at the updates to the website,and adjust all the scrapers.

Thank you For doing this! You’re amazing!

**Comment by mugol** [comment]
**Update:** I was able to get the mirror links again.

![mirrors2](https://user-images.githubusercontent.com/8999796/83320534-687c5d00-a1fd-11ea-9b15-bdb43894cef6.PNG)

I'm gonna to try to integrate the changes to the repo now. It looks like I just need to rewrite scrapers.mirrors and add an helper function to common.py to make a post request. Here's an example of what scrapers.mirrors will probably look like:

```
def mirrors(url):
    # find the player element and follow the embeded video link
    embeded_link = urljoin(base_url, _soup(url).find(id='iframeplayer')['src'])
    embeded_soup = _soup(embeded_link)
    script_list = embeded_soup.select('script')
    
    # get script elements
    script_texts = [se.getText() for se in embeded_soup.select('script')]

    #regex for the post request variables
    post_regex = re.compile('(?<=var VB_POST_URL = ")(.+?(?="))')
    token_regex = re.compile('(?<=var VB_TOKEN = ")(.+?(?="))')
    id_regex = re.compile('(?<=var VB_ID = ")(.+?(?="))')

    #put post request values in dictonary
    token_with_id = {}
    for s_element in script_list:
        s_text = str(s_element)
        post_match = post_regex.search(s_text)
        token_match = token_regex.search(s_text)
        id_match = id_regex.search(s_text)
        if post_match:
            token_with_id['post_url'] = post_match.group(1)
        if token_match:
            token_with_id['token'] = token_match.group(1)
        if id_match:
            token_with_id['id'] = id_match.group(1)
    
    # make post request to get encoded video info
    mirrors_with_links = mirrors_request(token_with_id)
    return mirrors_with_links
```

mirrors_request() will be the helper function in common.py, probably. I'll figure it out.

**Comment by aznhusband** [comment]
Great!  Gets us closer

**Comment by thebetauser** [comment]
@mugol is there a way to setup these python scripts to run in a pure python dev environment? If so, can you link me to any guides and I can help with this.

**Comment by mugol** [comment]
@thebetauser I think you _can_ technically run these scripts outside of Kodi if you want. However, your Python environment will probably throw a fit because the xbmc (or Kodi) library is not included. (I guess you can try using the Python environment in Kodi's files? I feel like that's not going to work as I imagine it would.)

Sorry I'm not much help in that regard. If it helps, here's the way I've been debugging this:

Editing the "production" files in the in the AppData folder (Not sure what this is in Mac/Linux):
`C:\Users\{your user here}\AppData\Roaming\Kodi\addons\plugin.video.icdrama\lib`

![thegoods](https://user-images.githubusercontent.com/8999796/83369856-769fba00-a372-11ea-9813-96b89a6c8d5f.png)


The files for the current repo should look like what's above. Kodi auto compiles the addon scripts every time there's a change, so you can just back out of the Icdrama addon and relaunch it in Kodi. Your changes should be live at that point. (Note: sometimes if the auto compile doesn't happen. I'd just relaunch Kodi.)

At this point you probably want logging, so you can add this anywhere in the script that has access to the xbmc library:
`xbmc.log('YOUR TESTING STRING HERE',level=xbmc.LOGNOTICE)`
![mirrorsdebug](https://user-images.githubusercontent.com/8999796/83370069-1d845600-a373-11ea-9300-e54ab051f588.png)
For example, I put a log message in actions.py's _mirrors function.

The example xbmc.log has the flag "LOGNOTICE". This will output to the txt file here:
`C:\Users\{your user here}\AppData\Roaming\Kodi\kodi.txt`
![logagain2](https://user-images.githubusercontent.com/8999796/83370595-bbc4eb80-a374-11ea-9d9d-8d6a032269da.png)



There are different levels of logging that I haven't really looked into. Here's a forum about them. Look at the third post:
https://forum.kodi.tv/showthread.php?tid=324570

That might have been a lot to digest for one comment. Let me know if you have any questions.

**Comment by aznhusband** [comment]
Made a bit of progress last night, still a long way to go.  Would be nice to have some kind of a test-jig setup so I could do the whole thing from the command line.

**Comment by jcheong-1** [comment]
Greatly appreciate everyone's efforts!

**Comment by chimin** [comment]
I have tried to run the resolveurl from standalone python, but since the website is broken now I can't sure I'm doing it correctly. Maybe you can see if this help? https://github.com/chimin/aznhusband.github.io

**Comment by dumbo0001** [comment]
@mugol @aznhusband This commit https://github.com/dumbo0001/aznhusband.github.io/commit/c4ffa3e207b89b6760a8e4089a819bc9102a5bfc seems to fix the 1st picasaweb/google video source. No resolving is needed for the scraped video URL. Hope this helps..

**Comment by thebetauser** [comment]
I can confirm that @dumbo0001 codesnippet was able to play the googlevideo mirrors

**Comment by aznhusband** [comment]
OK, we've made some breakthoughs, another user also send me a snippet that was more elegant than what I was working on.  I'll incorporate these changes, push a new version tonight.

**Comment by tofufi** [comment]
eager to see.......! TKS all effort 
    On Sunday, June 7, 2020, 01:22:15 p.m. EDT, aznhusband <notifications@github.com> wrote:  
 
 


OK, we've made some breakthoughs, another user also send me a snippet that was more elegant than what I was working on. I'll incorporate these changes, push a new version tonight.

**Comment by LQQK324** [comment]
I hope everything works out I missed my drama

**Comment by aznhusband** [comment]
OK, just pushed ICDrama 2.1.0.  Thanks to mugol, dumbo0001, and s_h (anonymous contributor) for the help.  Web scraping ain't my thing, and I was suffering.

I'll leave this open for a couple of days, just in case we encounter new bugs.

**Comment by jcheong-1** [comment]
OMG thanks aznhusband, mugol, dumbo0001, and s_h (anonymous contributor) for all of your hard work on maintaining this, really really appreciate everyone's efforts!

**Comment by antheroo** [comment]
Hi @aznhusband , not sure if it's due to the resolveurl script version I am using, but the unwrapped_url that my video resolved to was using 'googleusercontent.com'. Thus I had to add that check to get the plugin to work.

In file plugin.video.icdrama/lib/resolvers/icdrama.py, in get_media_url add check for '**googleusercontent.com**'

```
if ('googleusercontent.com' in unwrapped_url or
	'redirector.googlevideo.com' in unwrapped_url or
	'blogspot.com' in unwrapped_url or
	'fbcdn.net' in unwrapped_url): #for current Videobug source
	# Kodi can play directly, skip further resolve
	return unwrapped_url
```

**Comment by Ero-Sennin-Jiraiya** [comment]
Thank you to all the Contributors! Now we can all sit back and relax while the wife binwatch 
ICDrama 2.1.0   Thx again, a round of beer for all!

**Comment by LQQK324** [comment]
Finally I can watch my dramas. Thank you to all of you. Next round of beer
is on me oo

**Comment by LQQK324** [comment]
So far everything seems fine, not sure about others. But sometimes I get
the url limit exceeded or some of the video gets an error then hour later
it works.

**Comment by klingon888** [comment]
Unfortunately, for me, some shows are working but some are not. I've tried uninstalling and re-install, same prob. Below is the kodi log:

2020-06-09 10:01:35.184 T:11106  NOTICE: ResolveURL: Initializing ResolveURL version: 5.0.38
2020-06-09 10:01:47.376 T:11172  NOTICE: Previous line repeats 2 times.
2020-06-09 10:01:47.376 T:11172 WARNING: addons must not activate DialogBusy
2020-06-09 10:01:51.435 T:11172   ERROR: ResolveURL: UnboundLocalError Error - From: googlevideo Link: https://lh3.googleusercontent.com/-DslvnicTEEk/Xt-TGA6AwoI/AAAAAAAAAPs/kCKsAWDozNQHzZIo6YeaXBzVuWBxB5bvQCLcBGAsYHQ/m18/159171038239686917_la01.mp4?access_token=ya29.c.KpABzgfe0LEg-csS6yHwBER_SHW4KnXFLAo5hRMURuETqiGZm0_cbepogHs04YLTPyzCTW04vhR-FqeFrDy8bQ4-U5tAukOaf0ao3l78dm_yq46Lz1lp9aJglrhmXvm3D4yAoXupO-RNVvMvmtWF_SMMkCFJXMG-52amahQ_oa354Q6WZH0ye6jdFuT2_FIktEey: local variable 'response' referenced before assignment
2020-06-09 10:01:51.439 T:11172   ERROR: ResolveURL: UnboundLocalError Error - From: Icdrama Link: http://icdrama.to/vidembed.KjaDZ774bCjdRsY_s5nX4Q.html: local variable 'response' referenced before assignment
2020-06-09 10:01:51.461 T:11172   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.UnboundLocalError'>
                                            Error Contents: local variable 'response' referenced before assignment
                                            Traceback (most recent call last):
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.icdrama/addon.py", line 13, in <module>
                                                action_func(**kargs)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.icdrama/lib/actions.py", line 28, in make_dir
                                                diritems = func(*args, **kargs)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.icdrama/lib/actions.py", line 189, in mirrors
                                                return _mirrors(url)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.icdrama/lib/actions.py", line 204, in _mirrors
                                                play_mirror(url)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.icdrama/lib/actions.py", line 173, in play_mirror
                                                vidurl = common.resolve(iframe_url)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.icdrama/lib/common.py", line 84, in resolve
                                                return hmf.resolve()
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.module.resolveurl/lib/resolveurl/hmf.py", line 186, in resolve
                                                stream_url = resolver.get_media_url(self._host, self._media_id)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.icdrama/lib/resolvers/icdrama.py", line 45, in get_media_url
                                                return resolveurl.resolve(unwrapped_url)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.module.resolveurl/lib/resolveurl/__init__.py", line 131, in resolve
                                                return source.resolve()
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.module.resolveurl/lib/resolveurl/hmf.py", line 186, in resolve
                                                stream_url = resolver.get_media_url(self._host, self._media_id)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.module.resolveurl/lib/resolveurl/plugins/googlevideo.py", line 75, in get_media_url
                                                video = self._parse_redirect(web_url, hdrs=self.headers)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.module.resolveurl/lib/resolveurl/plugins/googlevideo.py", line 107, in _parse_redirect
                                                response_headers = dict([(item[0].title(), item[1]) for item in response.info().items()])
                                            UnboundLocalError: local variable 'response' referenced before assignment
                                            -->End of Python script error report<--
2020-06-09 10:01:51.887 T:11171   ERROR: GetDirectory - Error getting plugin://plugin.video.icdrama/?action=mirrors&url=http%3a%2f%2fwww1.adrama.to%2fwatch-online-life-after-death-episode-02-238893.html
2020-06-09 10:01:51.900 T:7344   ERROR: CGUIMediaWindow::GetDirectory(plugin://plugin.video.icdrama/?action=mirrors&url=http%3a%2f%2fwww1.adrama.to%2fwatch-online-life-after-death-episode-02-238893.html) failed

**Comment by Starz2play** [comment]
so far so good but I'm getting only 360P on piscaweb and the other available links (HLS, Hydrax and Uptobox) never work for me.  Is there a reason why?

**Comment by jcheong-1** [comment]
The addon just pulls the data from the website, the website only has 360p source and hence why you're only getting the 360p. 

Before reporting an issue, always best to check the site directly first to see first of all if the video is working, then check which sources are available.

**Comment by LQQK324** [comment]
Everything works fine for me. The only time it doesn't work is if the
website is not working properly. If you only get 360p and not 720p it's not
on the site yet.

**Comment by antheroo** [comment]
@Starz2play check the kodi logs and see. For my case the logs were showing CA Certificate errors. After updating the certificates then it works. Really depends on your device. Better to check the kodi.log

Sample error:
`2020-06-08 10:49:40.897 T:4066025472   ERROR: CCurlFile::Stat - Failed: Peer certificate cannot be authenticated with given CA certificates(60) for https://www46.uptostream.com`

**Comment by nilee8** [comment]
Hi there, Picasa links are all giving unresolvable address errors - 3 different devices all the same error:

2020-06-09 23:22:53.011 T:2328 WARNING: ResolveURL: Stream UrlOpen Failed: Url: https://lh3.googleusercontent.com/-yVfUqCO0Nx8/Xt-nxvAZpaI/AAAAAAAAAaM/pWQI4WFWJJYyYCMaGUfcrt-d2utekCFWACLcBGAsYHQ/m15/ez5c5xhc2ts2zznb2odik62abeiv2xb7brtjf45enhty5zxjzqeurp4m.source.mp4?access_token=ya29.c.KpABzgcWAh76s9IZEQpqbhTubNs7iTDDCcu9vuC1F10rZ3XUGmiNdlawqXRwvpxJa-hotLGGfDcCDiz6ueGV20a_1bESnQePj0Ry3aXSRBSSHBRTfmxSInfEkQ-ncGc7x8mmo3SWx-HMJXh-8qKcw3ij-KYeoEf5c2NDK1K7pI3NDJe-e8X7XsER82bvNy-5JuPB HTTP Code: 600 Msg: [Errno 11004] getaddrinfo failed
2020-06-09 23:22:53.062 T:1576   ERROR: XFILE::CDirectory::GetDirectory - Error getting plugin://plugin.video.icdrama/?action=mirrors&url=http%3a%2f%2fwww1.adrama.to%2fwatch-online-lo-and-behold-episode-900-238709.html
2020-06-09 23:22:53.074 T:10684   ERROR: CGUIMediaWindow::GetDirectory(plugin://plugin.video.icdrama/?action=mirrors&url=http%3a%2f%2fwww1.adrama.to%2fwatch-online-lo-and-behold-episode-900-238709.html) failed

**Comment by f33dm3** [comment]
Shit, i just got my wife back for 2 weeks.

~Thanks~

**Comment by chuachunmin** [comment]
Fresh install of kodi & icdrama works on MacOS.

I tried changing config.py on my Nvidia Shield from http://www.icdrama.to to adrama.to. Can access menu but can't access the video links(picasa etc). I tried changing current_ua in resolveURL settings.xml. Doesn't work.

**Comment by zmingxie** [comment]
> Fresh install of kodi & icdrama works on MacOS.
> 
> I tried changing config.py on my Nvidia Shield from http://www.icdrama.to to adrama.to. Can access menu but can't access the video links(picasa etc). I tried changing current_ua in resolveURL settings.xml. Doesn't work.
> 
> www1.icdrama.to is blocked.
> ![IMG_41BE180F4948-1](https://user-images.githubusercontent.com/26413860/86005210-e921b980-ba46-11ea-812c-fc2f39062e55.jpeg)
> 
> I fresh install kodi and icdrama on macbook. It didn’t work until I use Safari to load www1.adrama.to
> 
> On the Nvidia Shield, I load www1.adrama.to via the web viewer addon. Doesn’t work.

This is likely blocked by your ISP. Try to use Google DNS or Cloudflare and see if that helps.

**Comment by chuachunmin** [comment]
> > Fresh install of kodi & icdrama works on MacOS.
> > I tried changing config.py on my Nvidia Shield from http://www.icdrama.to to adrama.to. Can access menu but can't access the video links(picasa etc). I tried changing current_ua in resolveURL settings.xml. Doesn't work.
> > www1.icdrama.to is blocked.
> > ![IMG_41BE180F4948-1](https://user-images.githubusercontent.com/26413860/86005210-e921b980-ba46-11ea-812c-fc2f39062e55.jpeg)
> > I fresh install kodi and icdrama on macbook. It didn’t work until I use Safari to load www1.adrama.to
> > On the Nvidia Shield, I load www1.adrama.to via the web viewer addon. Doesn’t work.
> 
> This is likely blocked by your ISP. Try to use Google DNS or Cloudflare and see if that helps.

Thank you. That’s the solution.

**Comment by jcheong-1** [comment]
Looks like since a couple days ago the site has made some changes and it broke the current version from working.

reddit user kablamosayswhat posted a fix for 
Edit config.py Edit (line 6): domains = ['adrama.to', 'icdrama.to', 'icdrama.se'] to domains = ['adrama.to', 'icdrama.to', 'icdrama.se', 'hkdrama.to']

I've tested it and it worked for me on my box.

@aznhusband sorry I'm not a coder, would it be possible to update this and push a new update?

Thanks!

**Comment by Ero-Sennin-Jiraiya** [comment]
Saw the 2.2.0 update from https://aznhusband.github.io/plugin.video.icdrama/
confirmed (Line 6) updated and is working, thx all!
